### DISCUSSION TITLE
cpp/memory/pointer_declaration

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#37] [Announcements]
# cpp/memory/pointer_declaration

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/memory/pointer_declaration.html

**Comment by zyw0810** [comment]
1、先简化，不关注signal的参数，简化后：void (*signal())(int)；找到标识符signal，匹配()，表示是一个函数；*表示这个函数的返回值是一个指针，void (int)就是这个指针的类型，是一个参数是int返回值是void的函数的指针。
2、再看signal的参数，分别是int、void (*func)(int)，根据博主的推导方法，可以轻易知道void (*func)(int)是一个函数指针。综上，void (*signal(int sig, void (*func)(int)))(int)是一个函数，其返回值是void (*)(int)，其参数是int和void (*)(int)。
3、把影响分析的声明简化后再看，typedef void (*SB)(int)，套入void (*signal(int sig, void (*func)(int)))(int)中就是 SB signal(int, SB)。

**Comment by Falcom4000** [comment]
我建议忘记运算符优先级，采用Clockwise rule，参考https://c-faq.com/decl/spiral.anderson.html

**Comment by zyw0810** [comment]
这是来自QQ邮箱的假期自动回复邮件。帅哥美女，你的邮件我已经收到!

**Comment by sjoeycookie** [comment]
读法从内层到外层，拆解定义从外到内

**Comment by zyw0810** [comment]
这是来自QQ邮箱的假期自动回复邮件。帅哥美女，你的邮件我已经收到!
